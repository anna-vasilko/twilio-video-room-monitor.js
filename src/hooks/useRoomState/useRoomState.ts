import { useEffect, useState } from 'react';
import useRoom from '../useRoom/useRoom';

type RoomStateType = 'disconnected' | 'connected' | 'reconnecting';

export default function useRoomState() {
  const room = useRoom();
  const [state, setState] = useState<RoomStateType>('disconnected');

  useEffect(() => {
    if (room) {
      const setRoomState = () => setState((room.state || 'disconnected') as RoomStateType);
      setRoomState();
      room.on('disconnected', setRoomState).on('reconnected', setRoomState).on('reconnecting', setRoomState);
      return () => {
        room.off('disconnected', setRoomState).off('reconnected', setRoomState).off('reconnecting', setRoomState);
      };
    } else {
      setState('disconnected');
    }
  }, [room]);

  return state;
}
