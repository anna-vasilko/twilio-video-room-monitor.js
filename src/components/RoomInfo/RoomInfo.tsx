import React from 'react';
import Datum from '../typography/Datum/Datum';
import Headline from '../typography/Headline/Headline';
import useDominantSpeaker from '../../hooks/useDominantSpeaker/useDominantSpeaker';
import useIsRecording from '../../hooks/useIsRecording/useIsRecording';
import useRoom from '../../hooks/useRoom/useRoom';
import useRoomState from '../../hooks/useRoomState/useRoomState';
import useStats from '../../hooks/useStats/useStats';
import { Accordion } from '../typography/Accordion/Accordion';
import NestedObject from '../typography/NestedObject/NestedObject';
import ArrayData from '../typography/ArrayData/ArrayData';

export default function RoomInfo() {
  const room = useRoom();
  const dominantSpeaker = useDominantSpeaker();
  const roomState = useRoomState();
  const isRecording = useIsRecording();
  const { currentReceivedBitrate, currentSentBitrate } = useStats();

  if (!room) return null;

  const {
    audio,
    automaticSubscription,
    bandwidthProfile,
    dscpTagging,
    enableDscp,
    iceServers,
    iceTransportPolicy,
    insights,
    maxAudioBitrate,
    maxVideoBitrate,
    name,
    networkQuality,
    region,
    preferredAudioCodecs,
    preferredVideoCodecs,
    loggerName,
    tracks,
    video,
  } = room._options;

  return (
    <div style={{ marginBottom: '1em' }}>
      <Headline>Room information:</Headline>
      {roomState === 'disconnected' ? (
        <span>Not connected to a Twilio Video room.</span>
      ) : (
        <>
          <Datum label="Room Name" value={room.name} />
          <Datum label="SID" value={room.sid} />
          <Datum label="State" value={roomState} />
          <Datum label="Dominant Speaker" value={String(dominantSpeaker?.identity || null)} />
          <Datum label="Media Region" value={room.mediaRegion} />
          <Datum label="Is Recording" value={isRecording} />
          <Datum label="Total Sent Bandwidth" value={currentSentBitrate?.toLocaleString() + 'kbps'} />
          <Datum label="Total Received Bandwidth" value={currentReceivedBitrate?.toLocaleString() + 'kbps'} />

          <Accordion label="Connection Options">
            <NestedObject label="Audio" obj={audio} />
            <Datum label="Automatic Subscription" value={automaticSubscription} />
            <NestedObject label="Bandwidth Profile" obj={bandwidthProfile} />
            <Datum label="Dominant Speaker" value={room._options.dominantSpeaker} />
            <Datum label="DSCP Tagging" value={dscpTagging} />
            <Datum label="Enable DSCP" value={enableDscp} />
            <ArrayData label="Ice Servers" arr={iceServers} />
            <NestedObject label="Ice Transport Policy" obj={iceTransportPolicy} />
            <Datum label="Insights" value={insights} />
            <Datum label="Max Audio Bitrate" value={maxAudioBitrate} />
            <Datum label="Max Video Bitrate" value={maxVideoBitrate} />
            <Datum label="Name" value={name} />
            <NestedObject label="Network Quality" obj={networkQuality} />
            <Datum label="Region" value={region} />
            <ArrayData label="Preferred Audio Codecs" arr={preferredAudioCodecs} />
            <ArrayData label="Preferred Video Codecs" arr={preferredVideoCodecs} />
            <Datum label="Logger Name" value={loggerName} />
            <ArrayData label="Tracks" arr={tracks} />
            <NestedObject label="Video" obj={video} />
          </Accordion>
        </>
      )}
    </div>
  );
}
