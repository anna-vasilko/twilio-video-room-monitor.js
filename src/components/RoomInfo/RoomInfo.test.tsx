import React from 'react';
import { shallow } from 'enzyme';
import RoomInfo from './RoomInfo';
import useRoomState from '../../hooks/useRoomState/useRoomState';
import { LocalAudioTrack, LocalVideoTrack } from 'twilio-video';

jest.mock('../../hooks/useDominantSpeaker/useDominantSpeaker', () => () => ({
  dominantSpeaker: { identity: 'testIdentity' },
}));
jest.mock('../../hooks/useIsRecording/useIsRecording', () => () => 'false');
jest.mock('../../hooks/useRoom/useRoom', () => () => ({
  name: 'test123',
  sid: 'XXXXXXXXXXXXX1234',
  mediaRegion: 'testRegion',
  _options: {
    audio: { logLevel: {}, loggerName: 'twilio-audio', name: 'testName', workaroundWebKitBug180748: true },
    automaticSubscription: true,
    bandwidthProfile: {
      video: {
        dominantSpeakerPriority: 'standard',
        maxSubscriptionBitrate: 10,
        clientTrackSwitchOffControl: 'auto',
        contentPreferencesMode: 'auto',
        maxTracks: 100,
        mode: 'grid',
        renderDimensions: {
          high: { width: 1080, height: 720 },
          low: { width: 100, height: 70 },
          standard: { width: 340, height: 190 },
        },
        trackSwitchOffMode: 'predicted',
      },
    },
    dscpTagging: true,
    enableDscp: false,
    eventListener: null,
    iceServers: [],
    iceTransportPolicy: undefined,
    insights: true,
    maxAudioBitrate: 1600,
    maxVideoBitrate: null,
    name: 'testRoom',
    networkQuality: {
      local: 1,
      remote: 1,
    },
    region: 'testRegion',
    preferredAudioCodecs: [
      {
        isa: 'testIsa',
        opus: 'testOpus',
        PCMA: 'testPCMA',
        PCMU: 'testPCMU',
      },
    ],
    preferredVideoCodecs: [
      {
        codec: 'VP0',
        simulcast: false,
      },
      {
        codec: 'VP8',
        simulcast: true,
      },
    ],
    logLevel: {},
    loggerName: 'logger-twilio-video',
    tracks: [LocalAudioTrack, LocalVideoTrack],
    video: { logLevel: {}, loggerName: 'twilio-video', name: 'testName', workaroundWebKitBug180748: false },
  },
}));

jest.mock('../../hooks/useRoomState/useRoomState');
jest.mock('../../hooks/useStats/useStats', () => () => ({
  currentReceivedBitrate: 1234.56,
  currentSentBitrate: 6543.21,
}));

const mockUseRoomState = useRoomState as jest.Mock<any>;
mockUseRoomState.mockImplementation(() => 'connected');

describe('The RoomInfo component', () => {
  it('should render information about the room if connected to a Twilio Video room', () => {
    const wrapper = shallow(<RoomInfo />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should display warning when not connected to a Twilio Video room', () => {
    mockUseRoomState.mockImplementationOnce(() => 'disconnected');
    const wrapper = shallow(<RoomInfo />);
    expect(wrapper.contains(<span>Not connected to a Twilio Video room.</span>)).toBe(true);
  });
});
