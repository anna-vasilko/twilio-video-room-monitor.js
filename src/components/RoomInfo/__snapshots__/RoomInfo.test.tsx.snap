// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RoomInfo component should render information about the room if connected to a Twilio Video room 1`] = `
<div
  style={
    Object {
      "marginBottom": "1em",
    }
  }
>
  <Memo(styled.h3)>
    Room information:
  </Memo(styled.h3)>
  <Memo(Datum)
    label="Room Name"
    value="test123"
  />
  <Memo(Datum)
    label="SID"
    value="XXXXXXXXXXXXX1234"
  />
  <Memo(Datum)
    label="State"
    value="connected"
  />
  <Memo(Datum)
    label="Dominant Speaker"
    value={null}
  />
  <Memo(Datum)
    label="Media Region"
    value="testRegion"
  />
  <Memo(Datum)
    label="Is Recording"
    value="false"
  />
  <Memo(Datum)
    label="Total Sent Bandwidth"
    value="6,543.21kbps"
  />
  <Memo(Datum)
    label="Total Received Bandwidth"
    value="1,234.56kbps"
  />
  <Accordion
    label="Connection Options"
  >
    <Memo(NestedObject)
      label="Audio"
      obj={true}
    />
    <Memo(Datum)
      label="Automatic Subscription"
      value={true}
    />
    <Memo(NestedObject)
      label="Bandwidth Profile"
      obj="mockBandwidthProfile"
    />
    <Memo(Datum)
      label="Dominant Speaker"
      value={true}
    />
    <Memo(Datum)
      label="DSCP Tagging"
      value={true}
    />
    <Memo(Datum)
      label="Enable DSCP"
      value={false}
    />
    <Memo(ArrayData)
      arr={Array []}
      label="Ice Servers"
    />
    <Memo(NestedObject)
      label="Ice Transport Policy"
    />
    <Memo(Datum)
      label="Insights"
      value={true}
    />
    <Memo(Datum)
      label="Max Audio Bitrate"
      value={1600}
    />
    <Memo(Datum)
      label="Max Video Bitrate"
      value={null}
    />
    <Memo(Datum)
      label="Name"
      value="testRoom"
    />
    <Memo(NestedObject)
      label="Network Quality"
      obj={true}
    />
    <Memo(Datum)
      label="Region"
      value="testRegion"
    />
    <Memo(ArrayData)
      arr={
        Array [
          "PCMU",
        ]
      }
      label="Preferred Audio Codecs"
    />
    <Memo(ArrayData)
      arr={
        Array [
          "VP8",
        ]
      }
      label="Preferred Video Codecs"
    />
    <Memo(Datum)
      label="Logger Name"
      value="logger-twilio-video"
    />
    <Memo(ConnectionOptionTracks)
      tracks={
        Array [
          "mockTrack",
        ]
      }
    />
    <Memo(NestedObject)
      label="Video"
      obj={true}
    />
  </Accordion>
</div>
`;
